### DISCUSSION TITLE
新建多个线程，每个线程执行一个pag，为什么会崩溃，代码如下

### DISCUSSION THREAD

**Discussion by tang-1210** [Tencent/libpag#2602] [Q&A]
bool PagProcess::pagFrameProc()
    {

        // 等待音频和视频资源下载，初始化完成
        {
            std::unique_lock<std::mutex> lock(proc_mutex);
            proc_start_cv_.wait(lock, [this]
                            { return is_proc_flag_; });
        }

        //创建资源
        pag_surface_ = pag::PAGSurface::MakeOffscreen(pag_file_->width(), pag_file_->height());
        if (pag_surface_ == nullptr)
        {
            LError("session_id_ :{} open pag process failed pag_file_->width:{} pag_file_->height:{}", session_id_, pag_file_->width(), pag_file_->height());
            return false;
        }

        pag_player_ = new pag::PAGPlayer();
        pag_player_->setSurface(pag_surface_);
        pag_player_->setComposition(pag_file_);
        total_frame_num_ = floor(pag_file_->duration() * pag_file_->frameRate() / 1000000ll);
        frame_index_ = 0;
        bytes_len_ = pag_file_->width() * pag_file_->height() * 4;

        //计算步进
        if (pag_file_->frameRate() > target_fps_) {
            step_ = (float)pag_file_->frameRate() / target_fps_;
        }
        else if (pag_file_->frameRate() < target_fps_)
        {
            step_ = (float)target_fps_ / pag_file_->frameRate() - 1;
        }
        else 
        {
            step_ = 1;
        }

        if (pag_file_->frameRate() < target_fps_)
        {
            frameCounter += step_;        
        }
        else
        {
            frameDownCounter = step_;
        }

        LInfo("session_id_:{} pag_file_->width:{} pag_file_->height:{} total_frame_num_:{}", session_id_, pag_file_->width(), pag_file_->height(), total_frame_num_);
        while (is_proc_flag_)
        {
            // 判断是否需要重置
            if (frame_index_ > total_frame_num_) 
            {           
                LInfo("session_id_:{} reset frame_index_:{} total_frame_num_:{} total_image_num_:{} ",session_id_, frame_index_, total_frame_num_, total_image_num_);            
                break;    
            }
            pag_player_->setProgress(frame_index_ * 1.0 / total_frame_num_);
            auto status = pag_player_->flush();

            auto data = new uint8_t[bytes_len_];
            pag_surface_->readPixels(pag::ColorType::BGRA_8888, pag::AlphaType::Premultiplied, data,
                                pag_file_->width() * 4);

            present_image_ = cv::Mat(pag_file_->height(), pag_file_->width(), CV_8UC4, data).clone();

            /*缩放*/
            if (dst_width_ != pag_file_->width() || dst_height_ != pag_file_->height())
            {
                cv::resize(present_image_, present_image_, cv::Size(dst_width_, dst_height_));
            }

            std::string sid1 = path +"_"+ std::to_string(frame_index_) + ".png";

            cv::imwrite(sid1, present_image_);

            image_buf_.push_back(present_image_.clone());
                    
            delete[] data;
            frame_index_++;
            total_image_num_++;
            proc_read_cv_.notify_all();

            if (pag_file_->frameRate() >= target_fps_)
            {
                        
                    if (frame_index_ < frameDownCounter)
                    {
                        frame_index_++;
                    }

                    frameDownCounter += step_;
                }
                else
                {
                    if (frameCounter >= 1.0)
                    {
                        frame_index_--;
                        frameCounter = frameCounter - 1;
                    }
                    else
                    {
                        frameCounter += step_; 
                    }
                }

                //休眠1ms
                std::this_thread::sleep_for(std::chrono::milliseconds(1));
        }
        is_proc_flag_ = false;
        pag_surface_.reset();
        pag_file_.reset();
        delete pag_player_;
            
        return true;
    }

**Comment by kevingpqi123** [comment]
反馈问题建议将背景补充完整，如什么平台，采用的是什么渲染方案，基于 Angle 的渲染方案，好像多线程会有问题

**Comment by jinwuwu001** [comment]
该问题长时间未回复，先关闭，后续有问题可重新提
